url6a9f553762_Bresiu_socket_io_tgz-pJ8-MainJ8/src/com/github/nkzawa/engineio/client/transports/PollingXHR.java

```diff
@@ -0,0 +1,308 @@
+        private HttpURLConnection xhr;
+
+        public Request(Options opts) {
+            this.method = opts.method != null ? opts.method : "GET";
+            this.uri = opts.uri;
+            this.data = opts.data;
+            this.sslContext = opts.sslContext;
+        }
+
+        public void create() {
+            final Request self = this;
+            try {
+                logger.fine(String.format("xhr open %s: %s", this.method, this.uri));
+                URL url = new URL(this.uri);
+                xhr = (HttpURLConnection)url.openConnection();
+                xhr.setRequestMethod(this.method);
+            } catch (IOException e) {
+                this.onError(e);
+                return;
+            }
+
+            if (xhr instanceof HttpsURLConnection && this.sslContext != null) {
+                ((HttpsURLConnection)xhr).setSSLSocketFactory(this.sslContext.getSocketFactory());
+            }
+
+            Map<String, String> headers = new TreeMap<String, String>(String.CASE_INSENSITIVE_ORDER);
+
+            if ("POST".equals(this.method)) {
+                xhr.setDoOutput(true);
+                headers.put("Content-type", "application/octet-stream");
+            }
+
+            self.onRequestHeaders(headers);
+            for (Map.Entry<String, String> header : headers.entrySet()) {
+                xhr.setRequestProperty(header.getKey(), header.getValue());
+            }
+
+            logger.fine(String.format("sending xhr with url %s | data %s", this.uri, this.data));
+            xhrService.submit(new Runnable() {
+                @Override
+                public void run() {
+                    OutputStream output = null;
+                    InputStream input = null;
+                    BufferedReader reader = null;
+                    try {
+                        if (self.data != null) {
+                            xhr.setFixedLengthStreamingMode(self.data.length);
+                            output = new BufferedOutputStream(xhr.getOutputStream());
+                            output.write(self.data);
+                            output.flush();
+                        }
+
+                        Map<String, String> headers = new TreeMap<String, String>(String.CASE_INSENSITIVE_ORDER);
+                        for (String key : xhr.getHeaderFields().keySet()) {
+                            if (key == null) continue;
+                            headers.put(key, xhr.getHeaderField(key));
+                        }
+                        self.onResponseHeaders(headers);
+
+                        final int statusCode = xhr.getResponseCode();
+                        if (HttpURLConnection.HTTP_OK == statusCode) {
+                            String contentType = xhr.getContentType();
+                            if ("application/octet-stream".equalsIgnoreCase(contentType)) {
+                                input = new BufferedInputStream(xhr.getInputStream());
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

Use a standard utility function such as Apache Commons IO `IOUtils.toString(InputStream, Charset)` to read data from an input stream. Using a standard utility function makes your code easier to read and less prone to resource leaks. Do not write your own code to read data from the input stream. See the following diff on GitHub that makes this code change: [diff1](https://github.com/MyRobotLab/myrobotlab/commit/a4db462136ebf6c9cb4bc1e60bfac6d6f37a4b6e#diff-a9871318c9e95407fbecd277bb0f4729724c94216d5d5d57d13f743058b9c76dL242)

Learn more: [IOUtils.toString in the Apache Commons IO 2.11.0 API](https://commons.apache.org/proper/commons-io/apidocs/org/apache/commons/io/IOUtils.html#toString-java.io.InputStream-java.nio.charset.Charset-)